src/ContainerGetCapableTrait.php

```diff
@@ -0,0 +1,178 @@
+<?php
+
+namespace Dhii\Data\Container;
+
+use ArrayAccess;
+use Dhii\Util\String\StringableInterface as Stringable;
+use Exception as RootException;
+use InvalidArgumentException;
+use Iterator;
+use OutOfRangeException;
+use Psr\Container\ContainerExceptionInterface;
+use Psr\Container\ContainerInterface;
+use Psr\Container\NotFoundExceptionInterface;
+use Traversable;
+
+/**
+ * Common functionality for reading data from data sets.
+ *
+ * Supported data sets are:
+ * * Arrays
+ * * Objects
+ * * {@link \Psr\Container\ContainerInterface}
+ * * {@link \ArrayAccess}
+ * * {@link \Traversable}
+ *
+ * All SPL data storage classes are supported, given that they almost all implement either {@link \ArrayAccess} or
+ * {@link \Traversable}.
+ *
+ * @since [*next-version*]
+ */
+trait ContainerGetCapableTrait
+{
+    /**
+     * Retrieves a value from a container or data set.
+     *
+     * @since [*next-version*]
+     *
+     * @param array|object|Traversable|ContainerInterface $container The container to read from.
+     * @param string|Stringable                           $key       The key of hte value to retrieve.
+     *
+     * @return mixed The value mapped to the given key.
+     *
+     * @throws ContainerExceptionInterface If an error occurred while reading from the container.
+     * @throws NotFoundExceptionInterface  If the key was not found in the container.
+     */
+    protected function _containerGet($container, $key)
+    {
+        $key = $this->_normalizeString($key);
+        $isContainer = $container instanceof ContainerInterface;
+        $isArrayLike = is_array($container) || $container instanceof ArrayAccess;
+        $isObject = is_object($container);
+        $isTraversable = $container instanceof Traversable;
+
+        if (!$isContainer && !$isArrayLike && !$isObject && !$isTraversable) {
+            throw $this->_createInvalidArgumentException(
+                $this->__('Argument #1 is not a valid container'),
+                null,
+                null,
+                $container
+            );
+        }
+
+        if ($isContainer) {
+            return $container->get($key);
+        }
+
+        if ($isArrayLike && isset($container[$key])) {
+            return $container[$key];
+        }
+
+        if ($isObject && property_exists($container, $key)) {
+            return $container->{$key};
+        }
+
+        if ($isTraversable) {
+            try {
+                $iterator = $this->_resolveIterator($container);
+                foreach ($iterator as $_key => $_value) {
+                    if ($_key === $key) {
+                        return $_value;
+                    }
+                }
+            } catch (OutOfRangeException $outOfRangeException) {
+                // Do nothing. The iterator was not resolved, so the key was not found.
+            }
+        }
+
+        throw $this->_createNotFoundException($this->__('Key "%s" was not found', [$key]), null, null, null, $key);
```

Comment:
No need to put the key into the description. It's already part of the exception's data.

Reply:
True, but it adds meaning to the exception message, which is _meant_ to be meaningful.

Reply:
Without this, the description already says what is wrong. **With** what - that's the job of [`DataKeyAwareInterface#getDataKey()`].

[`DataKeyAwareInterface#getDataKey()`]:  https://github.com/Dhii/data-container-interface/blob/develop/src/DataKeyAwareInterface.php#L21